### PR TITLE
[#23] Add example user interface implementation

### LINKED ISSUE #23: Add example User Interface
This issue is part of the adaption process in order to qualify Hono as a replacement for Google IoT Core customers. We want to give users the possibility to operate with the Device Registry API as well as with the Device Communication API [#19]  with the help of an user interface.

In order to provide an user interface, frontend implementation based on Angular 15 has been made. This example implementation provides frontend functionality for a Google Cloud based deployment. We are recommending using [IAP](https://cloud.google.com/iap/docs) for your API security. Make sure to configure your [OAuth consent](https://developers.google.com/workspace/guides/configure-oauth-consent) screen. This UI will provide the possibility to:

_1. Access the Device Registry API containing actions like:_ 
- list tenants
- create a new tenant with an ID and messaging-type
- delete a tenant
- update a tenant with another messaging-type
- list devices of a tenant
- create a new device with an ID
- delete a device
- list credentials of a device
- create [Username/Password](https://www.eclipse.org/hono/docs/concepts/device-identity/#usernamepassword-based-authentication) or [JSON Web Token](https://www.eclipse.org/hono/docs/concepts/device-identity/#json-web-token-based-authentication) based credentials
- update JSON Web Token based credentials
- delete credentials

_2. Access the Device Communication API containing actions like:_
- list configs
- update a config 
- list states
- send a command

**Note**: This UI cannot be run without further adjustments! If one wants to use this UI in other environments than on Google Cloud, adjustments have to be made to not include the GoogleService and to update the url suffixes of the services. One consideration is to adjust the environment files in the future.

### PR DESCRIPTION
**Pull request by michelleFranke**
At this PR we have created an example user interface for a Google Cloud based deployment. The UI is able to communicate with the Device Registry API and the Device Communication API.

- Added new module to support a user interface based on Angular 15
- Added components for tenants, devices, credentials, configs and states
- Added services for communicating with the APIs
- Added Dockerfile
- Updated Readme

### CODE REVIEW

**Comment by calohmn** [comment]
@michelleFranke Thanks for the PR, this looks very nice and very useful!
(I had done a local test, commenting the `private googleService: GoogleService` constructor parameter in `api.service.ts` and `app.components.ts` to skip the Google login.)

In fact, it would be good to have such a device-management UI in the Hono repository itself, being more prominent there.
For that however, I think the Google service integration and the usage of the Device Communication API would need to be made optional. What do you think about such changes?
An option could also be to first merge this PR here in Hono-extras like it is, then develop the above mentioned changes and, once done, move it to the Hono repository.

Another thing: could you add a copyright header to the source files (mainly the typescript files)? 
The usual format would be like this:
```
/********************************************************************************
 * Copyright (c) 2023 Contributors to the Eclipse Foundation
 *
 * See the NOTICE file(s) distributed with this work for additional
 * information regarding copyright ownership.
 *
 * This program and the accompanying materials are made available under the
 * terms of the Eclipse Public License 2.0 which is available at
 * http://www.eclipse.org/legal/epl-2.0
 *
 * SPDX-License-Identifier: EPL-2.0
 ********************************************************************************/
```

**Comment by sanguyen123** [comment]
Good morning @calohmn I took over this frontend PR from michelleFranke. The option you mentioned to first merge this PR into Hono-extras sounds good. Regarding the above mentioned changes to the optional google service integration, the optional device communication API and the moving of the UI to the Hono repository itself is a good idea in our opinion, that we can discuss at a later point together.

I will add the copyright headers to the source files. There are also some improvements in usabilty and support for gateways has been added. Those changes will be commited soon and will be part of this PR.

**Comment by sanguyen123** [comment]
Hi @calohmn , I just pushed the recent changes, which mainly includes support for gateways, request-response commands, general UI improvements and including the header into the source files.

**Comment by calohmn** [comment]
@sanguyen123 Thanks for following up on this. As this is a quite large contribution, there is an IP check required. I've created a corresponding [ticket](https://gitlab.eclipse.org/eclipsefdn/emo-team/iplab/-/issues/10922). This will possibly take some days to be reviewed.

File: device-management-ui/src/assets/images/sort-up-solid.svg
Comment (calohmn) [inline device-management-ui/src/assets/images/sort-up-solid.svg:1]: @sanguyen123 There is a reference to a commercial license in these files:

device-management-ui/src/assets/images/sort-solid.svg
device-management-ui/src/assets/images/sort-up-solid.svg

Are these commercially licensed images?

**Comment by sanguyen123** [comment]
Hi @calohmn the two referenced SVGs fall under the free license of fontawesome and are open source / GPL friendly. I updated the SVGs to the newest svg-code from the offical website. Now it also states that it is Font Awesome Free. I hope this clarifies it.

For reference:

https://fontawesome.com/license/free
https://fontawesome.com/search?q=sort&o=r&m=free

**Comment by calohmn** [comment]
@sanguyen123 Thanks, the changes got approved in the IP check.

**Review by calohmn — APPROVED** [review]
LGTM, thanks a lot for the contribution!

**Comment by calohmn** [comment]
@sanguyen123 I've added #28 and #29 as follow-ups.

**Comment by julian-sotec** [comment]
Hi @calohmn - While it is great that the UI has been merged now - I have one little thing that I like to ask here:
We see that the commits have been squashed / merged together into one - which means the history of who did the changes / who is co author did not make it to the master. Is this something we can change after the merge? Any Ideas?

**Comment by calohmn** [comment]
@julian-sotec Sorry, as we usually squash PRs into one commit in Hono, I did so here as well, listing the individual authors in the commit message. If you prefer, I could add a commit that reverts the changes, so that you can then create a new PR based on the branch used here. That PR can then be completed without squashing.
One thing to note here though would be that, as discussed in the comments above, it would be good if we could move the UI code to the hono-repository after #28 and #29 have been resolved. Creating individual commits in the hono repo for each UI-related hono-extras commit would mean some effort then.

**Comment by julian-sotec** [comment]
> @julian-sotec Sorry, as we usually squash PRs into one commit in Hono, I did so here as well, listing the individual authors in the commit message. If you prefer, I could add a commit that reverts the changes, so that you can then create a new PR based on the branch used here. That PR can then be completed without squashing. One thing to note here though would be that, as discussed in the comments above, it would be good if we could move the UI code to the hono-repository after #28 and #29 have been resolved. Creating individual commits in the hono repo for each UI-related hono-extras commit would mean some effort then.

Hi @calohmn - thank you for clarifying this for us - I usually would expect to have only one commit - and ideally a small one. 
However since to us the history in this particular case is important We would prefere that you add a commit that reverts the changes so that we can complete the PR without squashing. Sorry for the additional effort.
To the second point when - I am afraid I have to ask what the additional effort with multiple commits when moving UI to hono would be in particular? Refering to "Creating individual commits in the hono repo for each UI-related hono-extras commit would mean some effort then."

**Comment by calohmn** [comment]
@julian-sotec The changes have been reverted now.

As for the 2nd point: My initial idea was, that just copying the UI files from the hono-extras repo to the Hono repo and then doing one commit there would be a lot easier than copying the commits somehow. But thinking more about it, the latter also seems rather straightforward (adding hono-extras repo as remote to the local hono repo, fetching, and then cherry-picking the commits with the UI changes).

**Comment by julian-sotec** [comment]
@calohmn sounds perfect to me, thanks!

**Pull request merged**